### ISSUE TITLE
Alignment Issue with reinterpret_cast on Embedded Systems

### ISSUE THREAD

**Issue by schkovich** [arobenko/cc.mqtt5.libs#3]
Hello, there! 

I encountered a `-Werror=cast-align` compiler error while building the library for an embedded system (RP2040 Connect) using the ARM toolchain (arm-none-eabi). This error arises when attempting to cast a `CC_Mqtt5ClientHandle` to a `cc_mqtt5_client::ClientImpl*` in `bm_client.cpp`.

The issue appears related to potential alignment mismatches between the source and target types. Below is the related compiler output:

```bash
[ 85%] Building C object client/lib/CMakeFiles/cc_mqtt5_bm_client.dir/bm/bm_client.c.obj
cd /home/ubuntu/cc.mqtt5.libs/bm-build/client/lib && /usr/bin/arm-none-eabi-gcc  -I/home/ubuntu/cc.mqtt5.libs/client/lib/src -I/home/ubuntu/cc.mqtt5.libs/client/lib/include -I/home/ubuntu/cc.mqtt5.libs/bm-build/client/lib/bm -isystem /home/ubuntu/cc.mqtt5.libs/bm-build/install/include -march=armv6-m -mthumb -O3 -DNDEBUG -fPIC -MD -MT client/lib/CMakeFiles/cc_mqtt5_bm_client.dir/bm/bm_client.c.obj -MF CMakeFiles/cc_mqtt5_bm_client.dir/bm/bm_client.c.obj.d -o CMakeFiles/cc_mqtt5_bm_client.dir/bm/bm_client.c.obj -c /home/ubuntu/cc.mqtt5.libs/bm-build/client/lib/bm/bm_client.c
/home/ubuntu/cc.mqtt5.libs/bm-build/client/lib/bm/bm_client.cpp: In function 'cc_mqtt5_client::ClientImpl* {anonymous}::clientFromHandle(CC_Mqtt5ClientHandle)':
/home/ubuntu/cc.mqtt5.libs/bm-build/client/lib/bm/bm_client.cpp:33:12: error: cast from 'CC_Mqtt5ClientHandle' {aka 'CC_Mqtt5Client*'} to 'cc_mqtt5_client::ClientImpl*' increases required alignment of target type [-Werror=cast-align]
   33 |     return reinterpret_cast<cc_mqtt5_client::ClientImpl*>(handle);
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/ubuntu/cc.mqtt5.libs/bm-build/client/lib/bm/bm_client.cpp: In function 'cc_mqtt5_client::op::ConnectOp* {anonymous}::connectOpFromHandle(CC_Mqtt5ConnectHandle)':
/home/ubuntu/cc.mqtt5.libs/bm-build/client/lib/bm/bm_client.cpp:43:12: error: cast from 'CC_Mqtt5ConnectHandle' {aka 'CC_Mqtt5Connect*'} to 'cc_mqtt5_client::op::ConnectOp*' increases required alignment of target type [-Werror=cast-align]
   43 |     return reinterpret_cast<cc_mqtt5_client::op::ConnectOp*>(handle);
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/ubuntu/cc.mqtt5.libs/bm-build/client/lib/bm/bm_client.cpp: In function 'cc_mqtt5_client::op::DisconnectOp* {anonymous}::disconnectOpFromHandle(CC_Mqtt5DisconnectHandle)':
/home/ubuntu/cc.mqtt5.libs/bm-build/client/lib/bm/bm_client.cpp:53:12: error: cast from 'CC_Mqtt5DisconnectHandle' {aka 'CC_Mqtt5Disconnect*'} to 'cc_mqtt5_client::op::DisconnectOp*' increases required alignment of target type [-Werror=cast-align]
   53 |     return reinterpret_cast<cc_mqtt5_client::op::DisconnectOp*>(handle);
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/ubuntu/cc.mqtt5.libs/bm-build/client/lib/bm/bm_client.cpp: In function 'cc_mqtt5_client::op::SendOp* {anonymous}::sendOpFromHandle(CC_Mqtt5PublishHandle)':
/home/ubuntu/cc.mqtt5.libs/bm-build/client/lib/bm/bm_client.cpp:83:12: error: cast from 'CC_Mqtt5PublishHandle' {aka 'CC_Mqtt5Publish*'} to 'cc_mqtt5_client::op::SendOp*' increases required alignment of target type [-Werror=cast-align]
   83 |     return reinterpret_cast<cc_mqtt5_client::op::SendOp*>(handle);
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/ubuntu/cc.mqtt5.libs/bm-build/client/lib/bm/bm_client.cpp: In function 'cc_mqtt5_client::op::ReauthOp* {anonymous}::reauthOpFromHandle(CC_Mqtt5ReauthHandle)':
/home/ubuntu/cc.mqtt5.libs/bm-build/client/lib/bm/bm_client.cpp:93:12: error: cast from 'CC_Mqtt5ReauthHandle' {aka 'CC_Mqtt5Reauth*'} to 'cc_mqtt5_client::op::ReauthOp*' increases required alignment of target type [-Werror=cast-align]
   93 |     return reinterpret_cast<cc_mqtt5_client::op::ReauthOp*>(handle);
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```

# Environment
## Host
**OS:** Ubuntu 24.04 Server
**CMake:** Version 3.28.3
**ARM Toolchain:** arm-none-eabi-gcc (Version 13.2.1)
## Target
**Board:** RP2040 Connect
**Architecture:** ARM Cortex-M0+ (ARMv6-M)
**Pointer Size:** 32-bit (CMAKE_SIZEOF_VOID_P=4)

# Build
## Toolchain File:
```cmake

set(CMAKE_SYSTEM "Generic")
set(CMAKE_SYSTEM_NAME "Generic")
set(CMAKE_SYSTEM_VERSION "")
set(CMAKE_SYSTEM_PROCESSOR "armv6-m")
set(CMAKE_SIZEOF_VOID_P "4")

set(CMAKE_CROSSCOMPILING "TRUE")

set(CMAKE_C_COMPILER /usr/bin/arm-none-eabi-gcc)
set(CMAKE_CXX_COMPILER /usr/bin/arm-none-eabi-g++)
set(CMAKE_ASM_COMPILER /usr/bin/arm-none-eabi-gcc)

set(CMAKE_C_COMPILER_WORKS TRUE CACHE BOOL "C Compiler works")
set(CMAKE_CXX_COMPILER_WORKS TRUE CACHE BOOL "C++ Compiler works")

message(STATUS "CMAKE_C_COMPILER=${CMAKE_C_COMPILER}")
message(STATUS "CMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}")

set(CMAKE_C_FLAGS "-march=armv6-m -mthumb" CACHE STRING "C flags")
set(CMAKE_CXX_FLAGS "-march=armv6-m -mthumb" CACHE STRING "C++ flags")
set(CMAKE_EXE_LINKER_FLAGS "--specs=nosys.specs -march=armv6-m -mthumb")

```

## Commands

```bash
#!/bin/bash

export ROOT_DIR="${HOME}/cc.mqtt5.libs"
export SCRIPT_DIR="${ROOT_DIR}/script"
export BUILD_DIR="${ROOT_DIR}/bm-build"
export COMMS_INSTALL_DIR="${BUILD_DIR}/install/lib/LibComms/cmake/"
export CCG_INSTALL_DIR="${BUILD_DIR}/install/lib/cc_mqtt5/cmake/"
export COMMON_INSTALL_DIR=${BUILD_DIR}/install
export COMMON_BUILD_TYPE=Release
export EXTERNALS_DIR="${HOME}/externals"
export BUILD_TYPE=Release
export CMAKE_TOOLCHAIN_FILE="${ROOT_DIR}/arm-none-eabi-toolchain.cmake"

if [ -d "${BUILD_DIR}" ]; then
    rm -rf "${BUILD_DIR}"
fi

mkdir -p "${BUILD_DIR}"

"${SCRIPT_DIR}"/prepare_externals.sh

cd "${BUILD_DIR}" || exit

cmake --trace .. -DCMAKE_TOOLCHAIN_FILE="${CMAKE_TOOLCHAIN_FILE}" \
        -DLibComms_DIR="${COMMS_INSTALL_DIR}" \
        -Dcc_mqtt5_DIR="${CCG_INSTALL_DIR}" \
        -DCMAKE_INSTALL_PREFIX=${COMMON_INSTALL_DIR} \
        -DCC_MQTT5_CUSTOM_CLIENT_CONFIG_FILES="${ROOT_DIR}/client/lib/script/BareMetalTestConfig.cmake" \
        -DCC_MQTT5_CLIENT_DEFAULT_LIB=OFF \
        -DCMAKE_BUILD_TYPE=${BUILD_TYPE} \
        -DCC_MQTT5_USE_CCACHE=OFF \
        -DCC_MQTT5_CLIENT_LIB_FORCE_PIC=ON 

procs=$(nproc)

if [ -n "${procs}" ]; then
    procs_param="--parallel ${procs}"
fi

cmake --build "${BUILD_DIR}" --config ${COMMON_BUILD_TYPE} --target install "${procs_param}"
```

# Analysis
The error occurs in the following code snippet:
```cpp
inline cc_mqtt5_client::ClientImpl* clientFromHandle(CC_Mqtt5ClientHandle handle)
{
    return reinterpret_cast<cc_mqtt5_client::ClientImpl*>(handle);
}
```

`CC_Mqtt5ClientHandle` is a pointer to a `CC_Mqtt5Client` object:
```cpp
typedef struct CC_Mqtt5Client* CC_Mqtt5ClientHandle;
```

The warning is triggered because the compiler cannot guarantee that the memory backing `CC_Mqtt5ClientHandle` (i.e., a `CC_Mqtt5Client` object) meets the stricter alignment requirements of cc_mqtt5_client::ClientImpl (the cast target).
# Needed help:
## Could you:

- ensure alignment guarantees for CC_Mqtt5Client and cc_mqtt5_client::ClientImpl are verified during allocation, such as through `alignas` or aligned allocators
- guide on resolving this issue when building for platforms with stricter alignment requirements, such as embedded systems.
- suggest if a safer casting mechanism or additional runtime checks are appropriate here.

Let me know if I can provide further details.

**Comment by arobenko** [comment]
The `CC_Mqtt5Client` and similar are "dummy" empty structs, they are not really used. Their purpose is to hide the real implementation details. There won't be any misaligned accesses during runtime. Disable treating warnings as errors (`-DCC_MQTT5_WARN_AS_ERR=OFF`) is safe in this instance.

Said that, I updated the code on "develop" branch to enforce suggested alignment on these dummy structs. Please try it and see if it works for you. If not, I'll just suppress the warning in the compilation flags.

Once you consider all your issues resolved please let me know, I'll do an official "patch" release.

**Comment by schkovich** [comment]
Thanks, that did not work. I tried `-DCMAKE_COMPILE_WARNING_AS_ERROR=false` but that failed too. 

Anyway, it's getting late on this side of the Globe. I will continue in the morning.

**Comment by arobenko** [comment]
Sorry, I introduced changes to the [cc.mqttsn.libs](https://github.com/commschamp/cc.mqttsn.libs) instead of this project. Didn't get my coffee fix before introducing the changes. :)
Will do it to this project as well soon.

**Comment by arobenko** [comment]
Please retry now

**Comment by schkovich** [comment]
Sounds good! :+1: 
```bash
$ arm-none-eabi-ar t bm-build/install/lib/libcc_mqtt5_bm_client.a 
ConnectOp.cpp.obj
DisconnectOp.cpp.obj
KeepAliveOp.cpp.obj
Op.cpp.obj
ReauthOp.cpp.obj
RecvOp.cpp.obj
SendOp.cpp.obj
SubscribeOp.cpp.obj
UnsubscribeOp.cpp.obj
ClientImpl.cpp.obj
TimerMgr.cpp.obj
bm_client.cpp.obj
arm-none-eabi-readelf -h bm-build/install/lib/libcc_mqtt5_bm_client.a 

File: bm-build/install/lib/libcc_mqtt5_bm_client.a(ConnectOp.cpp.obj)
ELF Header:
  Magic:   7f 45 4c 46 01 01 01 00 00 00 00 00 00 00 00 00 
  Class:                             ELF32
  Data:                              2's complement, little endian
  Version:                           1 (current)
  OS/ABI:                            UNIX - System V
  ABI Version:                       0
  Type:                              REL (Relocatable file)
  Machine:                           ARM
  Version:                           0x1
  Entry point address:               0x0
  Start of program headers:          0 (bytes into file)
  Start of section headers:          149464 (bytes into file)
  Flags:                             0x5000000, Version5 EABI
  Size of this header:               52 (bytes)
  Size of program headers:           0 (bytes)
  Number of program headers:         0
  Size of section headers:           40 (bytes)
  Number of section headers:         376
  Section header string table index: 375
# ...
```
Thank you! :smile:

**Comment by arobenko** [comment]
It's worth mentioning that "prepare_externals.sh" is a script I use to help me with development. It's not there for "official" build. I would recommend try to use [cc.cmake](https://github.com/commschamp/cc.cmake) project instead where the appropriate library(ies) is enabled. It is expected to build and install all the dependencies as well.

**Comment by schkovich** [comment]
Thanks, I considered [cc.cmake](https://github.com/commschamp/cc.cmake) when you [mentioned](https://github.com/commschamp/cc.mqtt5.libs/issues/2#issuecomment-2254379246) it earlier this summer. However, the project I am working on does not use cmake but PlatformIO custom build tools. I have `cc.mqtt5.libs` as a submodule and [statically linked](https://community.platformio.org/t/pre-compiled-static-libraries/179). [prepare_externals.sh](https://github.com/commschamp/cc.mqtt5.libs/blob/master/script/prepare_externals.sh) suits my needs better. 


PlatformIO offers a way to gain necessary control over the [network stack](https://github.com/Networking-for-Arduino/ESPHost). It's a laboriously customised setup, and moving away from the PlatformIO ecosystem is not practical now. I already have a functional asynchronous TCP client that utilises LWiP callbacks and [async_context](https://www.raspberrypi.com/documentation/pico-sdk/high_level.html#group_pico_async_context) from the Pico SDK.

**Comment by arobenko** [comment]
Fix merged to master, use [v1.0.2](https://github.com/commschamp/cc.mqtt5.libs/releases/tag/v1.0.2) release from now on.
